<?xml version="1.0"?>

<robot xmlns:xacro="http://ros.org/wiki/xacro">

  <!--<xacro:include filename="$(find p2os_urdf)/defs/materials.xacro"/>-->

  <!-- Properties (Constants) -->
  <xacro:property name="M_PI" value="3.14159"/>
  <!-- where to find all meshes -->
<xacro:property name="mesh_directory" value="package://technical_project/meshes/p3dx_meshes"/>

  <!-- Right/Left Hubcap + Wheel -->
  <xacro:macro name="powerbot_wheel" params="suffix parent reflect">
    <link name="powerbot_${suffix}_wheel">
      <inertial>
	<mass value="5.25"/> 
		<origin xyz="0 0 0"/>
		<inertia ixx="0.012411765597" ixy="0" ixz="0"
		         iyy="0.015218160428" iyz="0" izz="0.011763977943"/>
	      </inertial>
		<visual>
			<origin xyz="0 0 0" rpy="0 0 0"/>
			<geometry>
				<mesh filename="${mesh_directory}/wheel.stl" />
			</geometry>
			<material name="WheelBlack">
				<color rgba="0.117 0.117 0.117 1"/>
			</material>
		</visual>
	<collision>
		<origin xyz="0 0 0" rpy="${-3.1415927/2} 0 0"/>
		<geometry>
		  <cylinder radius="0.133" length="0.065"/>
		</geometry><!--
			<geometry>
				<mesh filename="${mesh_directory}/wheel.stl" scale= "1.005 1.005 1.005"/>
			</geometry>-->
	</collision>
    </link>

    <joint name="base_${suffix}_hubcap_joint" type="fixed">
      <origin xyz="0 0 0" rpy="0 0 0"/>

      <parent link="powerbot_${suffix}_wheel"/>
      <child link="powerbot_${suffix}_hubcap"/>
    </joint>

    <link name="powerbot_${suffix}_hubcap">
	      <inertial>
		<mass value="0.08"/>  
		<origin xyz="0 0 0"/>
		<inertia ixx="0.012411765597" ixy="0" ixz="0"
		         iyy="0.015218160428" iyz="0" izz="0.011763977943"/>
	      </inertial>
		<visual>
			<origin xyz="0 0 0" rpy="0 0 0"/>
			<geometry>
				<mesh filename="${mesh_directory}/hubcap.stl"/>
			</geometry>
        <material name="HubcapYellow">
          <color rgba="1.0 0.811 0.151 1.0"/>
        </material>
      </visual>
    </link>

    <!-- The transmission tag is used only by ros_control -->
    <!--<transmission type="pr2_mechanism_model/SimpleTransmission" name="${parent}_${suffix}_wheel_trans">
        <actuator name="base_${suffix}_wheel_motor" />
        <joint name="base_${suffix}_wheel_joint" />
        <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
        <mechanicalReduction>${reflect * 624/35 * 80/19}</mechanicalReduction>
        </transmission>-->
        
        
       
    <joint name="base_${suffix}_wheel_joint" type="continuous">
      <axis xyz="0 1 0"/>
      <anchor xyz="0 0 0"/>
      <limit effort="100" velocity="100" />
      <joint_properties damping="0.0" friction="0.0" />
      <origin xyz="0.084 ${reflect*0.35} -0.07" rpy="0 0 0"/>
      <parent link="base_center"/>
      <child link="powerbot_${suffix}_wheel"/>
    </joint>


    <transmission name="${suffix}_wheel_trans">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="base_${suffix}_wheel_joint">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
      </joint>
      <actuator name="base_${suffix}_wheel_motor">
        <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
        <mechanicalReduction>${reflect * 624/35 * 80/19}</mechanicalReduction>
      </actuator>
	  </transmission>

    <gazebo reference="powerbot_${suffix}_hubcap">
      <material value="Gazebo/Yellow"/>
    </gazebo>

    <gazebo reference="powerbot_${suffix}_wheel">
      <material value="Gazebo/Black"/>
      <mu1>0.5</mu1>
      <mu2>50.0</mu2>
      <kp>100000000.0</kp>
      <kd>1.0</kd>
    </gazebo>


  </xacro:macro>

      


</robot>

